### ISSUE TITLE
sql: add descriptor dumps to sentry reports during validation issues

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#57436]
Sentry reports about failed descriptor validations are very unactionable. For example, see #57376.

Unactionable sentry reports are the worst. We should improve this situation by adding a zipped, anonymized descriptor payload to the report.

Sentry reports have a maximum payload size of 200 kilobytes, however, so we have to be judicious.

@knz, can you please explain how we can add an arbitrary payload to a sentry report at error report time?

**Comment by postamar** [comment]
Weren't we going to use something called event-based logging or something like that to deal with this?

**Comment by ajwerner** [comment]
I think event based logging is totally orthogonal to this. This is about crash reports having more info. At least, I don't think anybody else was thinking about the two being related.

**Comment by knz** [comment]
> can you please explain how we can add an arbitrary payload to a sentry report at error report time?

After #77525 is merged, one can simply create a custom error wrapper with some descriptor payload(s), then implement a `SafeFormatError` that includes these descriptors when formatting the error verbosely.

I would not recommend including all the descriptors in the error btw. Error objects flow through many places in the system and the last thing we want is significant memory usage in objects that are not accounted for...

**Comment by postamar** [comment]
Agreed. We'll want to redact descriptor protos to keep them below a constant upper bound in byte size, as should be the case for any user-facing message really. We can hope that most descriptors are in the low kilobyte range to begin with, which will make these sentry reports at least sometimes useful, which is better than what we have now (i.e. nothing).

Do we still see a lot of validation issues in sentry these days? I don't think we do. Let's get this done with a very crude descriptor proto truncation scheme and if validation issues suddenly start popping up again we can backport more clever truncation schemes if need be.

**Comment by ajwerner** [comment]
What would be cool would be to attach objects to errors and could have some method to encode and redact those objects if they hit a crash report. 

As much as I'd like this, it's a bit of a task, so moving to the backlog.

**Comment by postamar** [comment]
We need to do something here, at least dump the current descriptor undergoing validation.

**Comment by ajwerner** [comment]
One is give me the string representation as an error detail because that thing is redacted but only semi-structured. The other is give me the redacted proto.

**Comment by ajwerner** [comment]
Another note here is, if we've collected cross-reference descriptors, let's get those too.
